### ISSUE TITLE
SPProductUpdate fails for 2 server farm when the ShutdownServices is set to True

### ISSUE THREAD

**Issue by MarekSarad** [PowerShell/SharePointDsc#832]
I am trying to automate SP update process with the SharepointDSC. One one farm scenario (like DEV machine it works ok) but for a 2 farm scenario is fails during pausing of the SSA. 

The error is either Unable to connect to remote server, or 'Connection was closed'(the latter I had only once so not sure about it) I can easily recreate the first one. After I complete installation on box 1 and try to run the code (copied form DSC) to pause the search from Box 2 I receive the first error.

>       $searchSAs = Get-SPEnterpriseSearchServiceApplication
>             foreach ($searchSA in $searchSAs)
>             {
>                 $searchSA.Pause()
>             }
> Exception calling "Pause" with "0" argument(s): "Unable to connect to the remote server"
> At line:5 char:17
> +                 $searchSA.Pause()
> +                 ~~~~~~~~~~~~~~~~~
>     + CategoryInfo          : NotSpecified: (:) [], MethodInvocationException
>     + FullyQualifiedErrorId : WebException

Is it something I am doing wrong? Should the 'ShutdownServices' be specified only for the server 1 and others should have it to 'false'? 

```
Configuration SPUpdate
{

    param(
            [Parameter(Mandatory = $true)]
            [PSCredential]$SetupAccount, 

            [Parameter(Mandatory = $true)]
            [string] $CUPath,

            [Parameter(Mandatory = $true)]
            [string] $CUFile

       )
      


    #This import is required to avoid the warrning message
    Import-DscResource -ModuleName 'PSDesiredStateConfiguration'
    Import-DscResource -ModuleName SharePointDsc
    Import-DsCResource -ModuleName xPendingReboot
    $FullPathLocal = 'C:\Temp\'+$CUFile
    Node "localhost"
    {
         LocalConfigurationManager
    {
         RebootNodeIfNeeded = $True
    }

        File CUDirectoryCopy
        {
            Ensure = "Present"  # You can also set Ensure to "Absent"
            Type = "Directory" # Default is "File".
            Recurse = $true # Ensure presence of subdirectories, too #MS Probably not necessary in our case
            SourcePath = $CUPath  
            DestinationPath = "C:\Temp"
        }

        Log AfterCUDirectoryCopy
        {
            # The message below gets written to the Microsoft-Windows-Desired State Configuration/Analytic log
            Message = "Finished running the file resource with ID DirectoryCopy"
            DependsOn = "[File]CUDirectoryCopy" # This means run "DirectoryCopy" first.
        }


        SPProductUpdate InstallCU
            {
                SetupFile            = 'C:\Temp\'+$CUFile 
                ShutdownServices     = $true
                Ensure               = "Present"
                PsDscRunAsCredential = $SetupAccount
                DependsOn = "[File]CUDirectoryCopy"
            }

        Log AfterInstallCU
        {
            # The message below gets written to the Microsoft-Windows-Desired State Configuration/Analytic log
            Message = "Finished applying the CU on the server"
            DependsOn = "[SPProductUpdate]InstallCU" 
        }

        # Reboot if pending

        xPendingReboot RebootCheck1

        {

            Name = "RebootCheck1"

        }

        SPConfigWizard PSConfig
        {
            Ensure = "Present"
            PsDscRunAscredential = $SetupAccount
        }

        Log AfterPSCOnfig
        {
            # The message below gets written to the Microsoft-Windows-Desired State Configuration/Analytic log
            Message = "Finished applying the CU on the server"
            DependsOn = "[SPConfigWizard]PSConfig" 
        }
    }
}
```

Running Win2012 R2 

SP16 January 2017 CU. 

I am using SharepointDSC 2.2.0.0 and PS 5.1

**Comment by johlju** [comment]
@MarekSarad Can you check if `$searchSA.Resume()` is called? I think `$searchPaused -eq $true` will never return true? 🤔 
Easiest is just add `Write-Verbose -Message 'Resuming Search' -Verbose` before the line 487, and the do a run with verbose output. 

https://github.com/PowerShell/SharePointDsc/blob/1aaf0b6382fc0b1bc5f40abbb768a6e6563811d6/Modules/SharePointDsc/DSCResources/MSFT_SPProductUpdate/MSFT_SPProductUpdate.psm1#L479-L490

**Comment by MarekSarad** [comment]
@johlju, thanks for reply.  
I will check that but I don't think I ever get to the Resume section. 
The scenario is. 
1. Same DSC is applied to 2 servers 
2. DSC stops the search from server 1. Starts binary installation. 
3. DSC tries to stop search from server 2 and then it fails. 

it fails at this stage of the code:

```
  $searchSAs = Get-SPEnterpriseSearchServiceApplication
        foreach ($searchSA in $searchSAs)
        {
            $searchSA.Pause()
        }
```

I will add lines to Pause and Resume and post results bit later.

**Comment by johlju** [comment]
My thought was it should start Search on first node before running the configuration on the second node. The Search Application is per farm? So it can only be paused on one server, and should not be started before all the nodes has been updated. So my thought is that it should only be run on one node at a time? So the intended behavior must be that one node should have this property set to $true and the others to $false, but then how to make sure the first one is always run first? Probably the `BinaryInstallTime` is meant to make sure that the first node is always started first. But then how to make sure the other nodes have finished before restarting Search on the first node. 🤔  

@ykuijs can you confirm or correct me?

**Comment by johlju** [comment]
Agree that the error should be resolved either way.
 
Multiple starts and stops would be good to avid, but what I is worried about what (bad) impact it would have if Search is started during patching of other nodes. I'm out of my depth here, so leave it up to the community to find the best solution to this problem. 
Labeling this as a bug for the time being.

**Comment by ykuijs** [comment]
Hi @MarekSarad, since DSC is only aware of the state of the server it is running on, using Shutdownservices on all servers will cause a second server to try to pause the Search even though the first server has already done that.

I agree with your proposed solution as a first enhancement, but as a final solution we do have to think about that a little further. For example: What if server1 completes and resumes the search while server2 is still working on the update process?

We are thinking about a solution to keep track on the status of each server and prevent these types of issues. But this is a pretty challenging problem ;-) See issue #457 for more information.

**Comment by MarekSarad** [comment]
@johlju I am not sure would starting of the search have bad impact on the other nodes. As we stop services on all servers they would not respond to the crawler queries anyway correct? So in that case search should not affect the binary installation (and I think that part is most affected by search crawler).

@ykuijs I understand the complexity of the issue. The big issue in general is that DSC is lacking orchestration. I can easily work around it as I use Octopus to deliver DSC and I can do create new step to stop services deliver DSC and then start them. But with just native DSC this is not possible.

**Comment by ykuijs** [comment]
Fix will be included in my next Bugfix PR

**Comment by MarekSarad** [comment]
I am testing locally fix with adding the  if( $searchSA.isPaused() -eq 0){  $searchSA.Pause()} but I am getting mixed results. 

Again same scenario 1 APP server, 1 FE server. When executed on the APP server after the pause completes $searchSA.isPaused() returns 128

But when executed on the FE it errors. 

`PS C:\Users\_spinstall> $searchSAs = Get-SPEnterpriseSearchServiceApplication
PS C:\Users\_spinstall> $searchSAs.isPaused()
Exception calling "IsPaused" with "0" argument(s): "Unable to connect to the
remote server"
At line:1 char:1
+ $searchSAs.isPaused()
+ ~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : NotSpecified: (:) [], MethodInvocationException
    + FullyQualifiedErrorId : WebException
`
That tested in PS window. But the error in DSC execution is the same:

VERBOSE: [DFS-SP16FE-750]: LCM:  [ End    Set      ]  [[SPProductUpdate]InstallCU]  in 3.4580 seconds. 
July 11th 2018 11:14:33

Error
PowerShell DSC resource MSFT_SPProductUpdate  failed to execute Set-TargetResource functionality with error message:  

July 11th 2018 11:14:33

Error
Exception calling "IsPaused" with "0" argument(s): "Unable to connect to the remote server"  

July 11th 2018 11:14:33

Error
    + CategoryInfo          : InvalidOperation: (:) [], CimException 


July 11th 2018 11:14:33

Error
    + FullyQualifiedErrorId : ProviderOperationExecutionFailure 
July 11th 2018 11:14:33

Error
    + PSComputerName        : localhost

**Comment by MarekSarad** [comment]
I managed to workaround the issue of checking if the search is paused using this approach

`            $searchSAs = Get-SPEnterpriseSearchServiceApplication
            foreach ($searchSA in $searchSAs)
            {
            Write-Verbose -Message "Before catch on searach Pause"
            try{
                    if($searchSA.isPaused() -eq 0)
                    {Write-Verbose -Message "Pausing Search"



                     $searchSA.Pause()
                     }
                 }
                 catch
                 {
                    $ErrorMessage = $_.Exception.Message
                    $FailedItem = $_.Exception.ItemName
                    if ($ErrorMessage -like "*Unable to connect to the remote server*"){Write-Verbose -Message "Unable to connect to the server, search application probably is paused"}
                    else {throw $ErrorMessage}

                 }
            }`

**Comment by ykuijs** [comment]
Closed via PR #838
